src/dodal/common/crystal_metadata.py

```diff
@@ -0,0 +1,65 @@
+import math
+from dataclasses import dataclass
+from enum import Enum
+from typing import Literal
+
+
+@dataclass(frozen=True)
+class Material:
+    """
+    Class representing a crystalline material with a specific lattice parameter.
+    """
+
+    name: str
+    lattice_parameter: float  # Lattice parameter in meters
+
+
+class MaterialsEnum(Enum):
+    Si = Material(name="silicon", lattice_parameter=5.4310205e-10)
+    Ge = Material(name="germanium", lattice_parameter=5.6575e-10)
+
+
+@dataclass(frozen=True)
+class CrystalMetadata:
+    """
+    Metadata used in the NeXus format,
+    see https://manual.nexusformat.org/classes/base_classes/NXcrystal.html
+    """
+
+    usage: Literal["Bragg", "Laue"] | None = None
+    type: str | None = None
+    reflection: tuple[int, int, int] | None = None
+    d_spacing: tuple[float, str] | None = None
+
+    def __init__(
+        self,
+        material: MaterialsEnum,
+        reflection_plane: tuple[int, int, int],
+        usage: str = "Bragg",
+    ) -> None:
+        # Determine material from reflection plane prefix
+        # Set attributes using object.__setattr__ since the class is frozen
+        assert all(
+            i > 0 for i in reflection_plane
+        ), "Reflection plane indices must be positive integers"
+        object.__setattr__(self, "type", material.value.name)
+        object.__setattr__(self, "reflection", reflection_plane)
+        d_spacing = self.calculate_d_spacing(
+            material.value.lattice_parameter, reflection_plane
+        )
+        object.__setattr__(self, "d_spacing", d_spacing)
+        object.__setattr__(self, "usage", usage)
+
+    @staticmethod
+    def calculate_d_spacing(
+        lattice_parameter: float, reflection: tuple[int, int, int]
+    ) -> tuple[float, str]:
+        """
+        Calculates the d-spacing value in nanometers based on the given lattice parameter and reflection indices.
+        """
+        h_index, k_index, l_index = reflection
```

Comment:
Do we have sources for these values? I believe the values used up until now have ultimately been derived from literature as they are experimentally measured properties, and I think the feeling is that we should cite sources for these values and use them directly rather than calculate them.

Reply:
I heard from @iain-hall  that there is a formula for the d-spacing from the reflection parameters

if we're talking about a lattice parameter sure it's a good idea to include a link
https://physics.nist.gov/cgi-bin/cuu/Value?asil

Reply:
You should probably consult with a controls engineer to make sure the formula is general and not only applicable to I18's DCM, otherwise you could break I22's. The DCMs are all subtly different and were designed in blocks. @JamesOHeaDLS might know more.

Reply:
Hello,

There is a general formula for deriving d-spacing from a particular crystals unit cell length and miller indices (d = a/sqrt(h^2 + k^2 + l^2) but the unit cell length is temperature dependent, and at the motor controller level we specify the d-spacing, rather than derive it.

For our coordinate system calculations we take the d-spacing from literature.  For example the 'Accel DCM' coordinate system has, as a default, the d-spacing of Silicon 111 (3.1355 A).  This would have been taken from reference literature.

This coordinate system parameter is overridable, so if the beamline wanted to use a different value, or the mono in question uses a different type of crystal, it could be specified.  It is also settable and gettable from the EPICS layer.

For example there is the X-Ray Data Booklet (you see copies of this orange booklet all over DLS. I have a copy myself). 
 See section 4.1  - https://xdb.lbl.gov/Section4/Sec_4-1.html which tabulates d-spacings for a range of crystals.  Here Silicon 111 has a value of 6.2712 in its 2d column (ie a d of 3.1356).

Note that the stated d-spacing can vary with temperature.  The above is for room temperature.  It is worth noting that the Diamond quad crystal mono coordinate system uses a hard-coded d-spacing for Si111 at liquid nitrogen temperature (3.13475).

So I think it makes more sense to just get the beamline scientists to specify the d-spacing they wish to use, and then ensure that the software (at all levels) uses the same value.  

James

Reply:
>  It is also settable and gettable from the EPICS layer.

If this is true for DCMs on all the beamlines then I strongly support just always reading it from EPICS and using that as the one source of truth.

Reply:
Yes this is true for all dual crystal monos, but not for quad (but could be very easily)

Reply:
thanks @JamesOHeaDLS for the in-depth explanation.

now I agree with @DominicOram that EPICS is the best source.


Revising the PR now

Reply:
added the EPICS source, resolving the conversation